source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class LazyCompatibleInterface {
+public:
+  // Helper operators to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* operator->() PURE;
+  virtual StatsStructType& operator*() PURE;
+  virtual ~LazyCompatibleInterface() = default;
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ */
+template <typename StatsStructType>
+class LazyInit : public LazyCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
+      : inited_(Stats::Utility::gaugeFromElements(
+            *scope, {Stats::DynamicName(StatsStructType::typeName()), Stats::DynamicName("inited")},
+            Stats::Gauge::ImportMode::Accumulate)),
+        ctor_([scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          inited_.inc();
+          return new StatsStructType(stat_names, *scope);
+        }) {
+    if (inited_.value() > 0) {
```

Comment:
we can save some memory easily here not saving the ctor (or setting it to nullptr) when we are auto-initializing here.

Reply:
I am not sure how much that makes a difference: assinging to ctor_ is required in instantiate() unless you are suggesting two copies of the logic.

Reply:
I am suggesting having the only initializer to the ctor in the else block, not duplicating it. This saves the construction and subsequent cleanup of the ctor in the normal case where we are refreshing a live cluster.

Reply:
so how to save a "ctor_" that's trivial to construct, now: 
1. if inited_ > 0: we would copy the ctor_ content (the ctor_ body) and paste it in the if statement directly?
2. else: create the ctor_ (with the and wait until it's been used. 

Is this what you are suggesting? 

my thinking is that it doesn't worth the improvements sacrificing the readability.

Reply:
```
  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
      : inited_(Stats::Utility::gaugeFromElements(
            *scope, {Stats::DynamicName(StatsStructType::typeName()), Stats::DynamicName("inited")},
            Stats::Gauge::ImportMode::Accumulate)) {
    if (inited_.value() > 0) {
      instantiate();
    } else {
     ctor_ = [scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
          inited_.inc();
          return new StatsStructType(stat_names, *scope);
        };
```

Seems *more* readable to me, and more efficient.

Reply:
so in instantiate(), do we create the ctor_ , by C&P the code piece in the else branch, every time it's called?

Reply:
agreed; we should drop a TODO for this though (requires adding a setter to AtomicPtr).

Reply:
there is already a comment at line 90